### DISCUSSION TITLE
ssr 预渲染构建后浏览器访问空白

### DISCUSSION THREAD

**Discussion by helvenk** [umijs/umi#6000] [Q&A]
ssr构建后，用 `serve` 来本地部署，访问时先正常渲染出页面，再变成空白。

查看 F12 工具发现响应的 html 是正确的，确实做到了 ssr，但是在执行 js 后将页面替换为空白了。

调试代码发现 build 后的 umi.js 文件中存在路由映射，可能是这个原因导致的。

```js
[{ path:"/test.html", hideInMenu:!0, component: n("IRZr").default, exact:!0 }]
```
访问路径 /test.html 会正常渲染，但实际访问路径是 /test，似乎是没有匹配到对应组件，React 渲染 root 为空白。

上面是加了 `htmlSuffix: true` 后的效果，path 会自动加上 .html 后缀，导致不能正确渲染，似乎可以接受。

但现在的场景是静态文件会被放到服务器上，可能增加某些前缀，比如 /a/b/c/test，这样就永远匹配不上了，导致上述的 bug 出现。

希望给个解决方案。

umi 版本是 3.2，ssr 相关配置如下：
```js
{    
    dva: false,
    locale: false,
    dynamicImport: false,
    ssr: {},
    esbuild: false,
    model: false,
    initialState: false,
    publicPath: '/',
    outputPath: 'dist',
    exportStatic: {
      dynamicRoot: true,
      htmlSuffix: true,
    },
}
```
路由随便写都行
```js
{ path: 'test', component: './Test' }
```

**Comment by sorrycc** [comment]
> 但现在的场景是静态文件会被放到服务器上，可能增加某些前缀，比如 /a/b/c/test，这样就永远匹配不上了，导致上述的 bug 出现。

那得配 base 吧，参考文档里的 base 属性。

**Reply by helvenk** [reply]
ssr的目的就是生成 html 文件交给后端处理，他们放哪里、以什么路径去 response 都是跟前端无关的，所以配置 base 不适合。

比如后端的路由匹配：
以 /user/info 来响应 a.html
以 /admin/info 来响应 b.html

甚至我们 build 后的静态 html 都可能是后端的模板 html，比如 thymeleaf 之类的。

所以还是需要有个解决方案，尝试过使用 next.js 构建后是可以放置到任意路径的。

**Comment by sorrycc** [comment]
对 ssr 的理解可能有误？ssr 是后端给 html，但逻辑部分还是由前端渲染的。

**Reply by helvenk** [reply]
需求上来说就是希望 ssr 打包过后支持以任意路径访问，试过 dynamicRoot 是不行的。

我的业务场景是要生成静态 html 交给后端去部署渲染（部署到 server 容器中，不通过 nginx 管理），他们的路由前端是无法预知的。这些 html 逻辑较少，基本上是纯展示，但是有复杂的样式，需要用到 antd （组件化），不能直接纯手写。
